Sources/Aperture/Aperture.swift

```diff
@@ -1,228 +1,823 @@
+					streamConfig.height = Int(windowFilter.contentRect.height)
+				} else {
+					streamConfig.width = Int(window.frame.width)
+					streamConfig.height = Int(window.frame.height)
+				}
+
+				filter = windowFilter
+			case .externalDevice:
+				filter = nil
+			case .audioOnly:
+				guard let display = content.displays.first else {
+					throw Error.noDisplaysConnected
+				}
+
+				let screenFilter = SCContentFilter(display: display, excludingWindows: [])
+				filter = screenFilter
 			}
+
+			if let filter {
+				stream = SCStream(filter: filter, configuration: streamConfig, delegate: self)
+			}
+
+			if let stream {
+				do {
+					try stream.addStreamOutput(self, type: .screen, sampleHandlerQueue: .global())
+
+					if options.recordSystemAudio {
+						try stream.addStreamOutput(self, type: .audio, sampleHandlerQueue: .global())
+					}
+
+					if #available(macOS 15, *), options.microphoneDeviceID != nil  {
+						try stream.addStreamOutput(self, type: .microphone, sampleHandlerQueue: .global())
+					}
+				} catch {
+					throw Error.couldNotStartStream(error)
+				}
+			}
+
+			do {
+				try await initOutput(target: target, options: options, streamConfig: streamConfig)
+			} catch {
+				let finalError: Aperture.Error
+				if let error = error as? Aperture.Error {
+					finalError = error
+				} else {
+					finalError = Error.couldNotStartStream(error)
+				}
+
+				try? await cleanUp()
+				reset()
+				onError?(finalError)
+
+				throw finalError
+			}
+
+			onStart?()
 		}
 
-		if session.canAddInput(input) {
-			session.addInput(input)
-		} else {
-			throw Error.couldNotAddScreen
+		public func stopRecording() async throws {
+			if let error {
+				reset()
+				/// We do not clean up here, as we have already cleaned up when the error was recorded
+				throw error
+			}
+
+			try? await cleanUp()
+			reset()
+			onFinish?()
 		}
 
-		if session.canAddOutput(output) {
-			session.addOutput(output)
-		} else {
-			throw Error.couldNotAddOutput
+		private func recordError(error: Error) {
+			self.error = error
+			onError?(error)
+			Task { try? await self.cleanUp() }
 		}
 
-		if let videoCodec = videoCodec {
-			let connection = try output.connection(with: .video)
-				.unwrapOrThrow(Error.couldNotSetVideoCodec)
-			output.setOutputSettings([AVVideoCodecKey: videoCodec], for: connection)
+		private func reset() {
+			self.target = nil
+			self.options = nil
+
+			self.error = nil
+			self.continuation = nil
+			self.activity = nil
+
+			self.stream = nil
+			self.isStreamRecording = false
+
+			self.assetWriter = nil
+			self.videoInput = nil
+			self.systemAudioInput = nil
+			self.microphoneInput = nil
+
+			self.microphoneDataOutput = nil
+			self.externalDeviceDataOutput = nil
+
+			self.microphoneCaptureSession = nil
+			self.externalDeviceCaptureSession = nil
+
+			self.isRunning = false
+			self.isPaused = false
+			self.isResuming = false
+			self.timeOffset = CMTime.zero
+			self.lastFrame = nil
 		}
```

Comment:
Having to reset this much state is kinda fragile. It would be easy to introduce more state and forget to reset it here. This is usually a sign that it needs some better encapulation. For example, a wrapper class that initiates this class on start record and destroys it on stop record. Then there is no need to reset.

Reply:
Renamed this to `RecordingSession` and wrapped in a new `Recorder` class